### PR TITLE
Support for Units parsing and conversions

### PR DESCRIPTION
**Pull request by dschwen**
This is the initial bit of work to get started on unit parsing and conversions and implements a parser extension to allow for explicit unit conversion `${units 3.2 mJ/mol -> eV/at}`.

Refs #14827

### CODE REVIEW

File: framework/src/parser/Parser.C
Comment (rwcarlsen) [inline framework/src/parser/Parser.C]: Should this maybe go after the `replace` evaler?

**Review by rwcarlsen — COMMENTED** [review]
This looks good - and like it.  What do others in @idaholab/moose-team think?  Do you want to permanently introduce @dschwen's fancy unit nonsense to input files?

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
We should probably get a new or updated design doc and a corresponding `design=foo.md` in the test spec file.

File: framework/src/parser/Parser.C
Comment (rwcarlsen) [reply framework/src/parser/Parser.C]: On second thought - this probably doesn't matter - since evalers are all executed in a single sweep over the AST - rather than one sweep per evaler.

Comment (rwcarlsen) [reply framework/src/parser/Parser.C]: To be clear - over the brace expression AST inside every parameter - and they are evaluated depth-first.

**Comment by permcody** [comment]
Unit conversion is very tricky to get right. To be anywhere near the ballpark of foolproof one would have to use the unit system nearly every place any constant or function was used. This PR might get you pretty far down the path of handling conversions inside of InputParameters but unless there is either a canonical unit system for the application, or all access into the InputParameter systems use unit information we run the risk of missing unit information somewhere. There are also bigger stumbling blocks to consider. You might go through your input file and put units everywhere but what about your mesh? Are we going to propagate units through there too? 

I guess what is your end goal here? To convert input file parameters to a single canonical system across all apps (so we don't have to think about mesh and other internals) or are you hoping for a more complete overhaul of how constants are used everywhere? 

I worked on another system that did unit conversions and we propagated it through to every single part of the system. It was nearly impossible to access a value without accessing a unit with it. It was extremely flexible as you could be sloppy and use mixed units all of the place and your calculations would still come out correctly but it necessarily creeped it's way through the whole system.

**Comment by dschwen** [comment]
I'm tabling the idea of propagating units through to input parameters for now. 

This PR adds a parser function to perform conversions within input files. The expressions are basically unit conversions that return pure number (just like an `${fparse ...}` expression would).

This will probably ease 90% of our pain when getting parameters from the literature. Currently we do unit conversion of a piece of paper and plug in bare numbers. If you're lucky you get the units as a comment, if you are very lucky you get the original value and the math for teh conversion as a comment.

Having a dedicated unit conversion doohickey will make this much simpler IMO and more transparent / self documenting.

File: test/tests/parser/param_substitution/unit_conversion.i
Comment (YaqiWang) [inline test/tests/parser/param_substitution/unit_conversion.i]: Can things like `${units 1 km -> m}` be used to replace any constant in an input file?

Comment (rwcarlsen) [reply test/tests/parser/param_substitution/unit_conversion.i]: Yes - the `units` brace expression just evaluates to a  number.  After resolving it, the input file above effectively looks like `function = 1000`.

**Comment by aeslaughter** [comment]
Can we get a short description/presentation about this and its end use for the MOOSE developers?

**Comment by dschwen** [comment]
> Can we get a short description/presentation about this and its end use for the MOOSE developers?

Is [this (for developers)](https://mooseframework.inl.gov/docs/PRs/14845/site/source/utils/Units.html) and [this (for end users)](https://mooseframework.inl.gov/docs/PRs/14845/site/application_usage/input_syntax.html) not enough?

**Review by rwcarlsen — DISMISSED** [review]
Looks good to me.  @permcody - I think this is a good approach - it just allows convenience conversions/cals to occur at the input file level while keeping MOOSE internals entirely agnostic to units - which is something already possible in input files anyway with fparse like @dschwen mentioned.

**Review by lindsayad — DISMISSED** [review]
Impressive how fast you got this up to speed

**Comment by andrsd** [comment]
No Fahrenheit?! 😈

**Comment by rwcarlsen** [comment]
What's up with the relap-7 failures (@andrsd or @joshuahansel any ideas)?

**Comment by andrsd** [comment]
> What's up with the relap-7 failure

Looking... (@joshuahansel is on travel)

File: framework/src/utils/Units.C
Comment (andrsd) [inline framework/src/utils/Units.C]: probably 'hour', rather than 'day'

Comment (andrsd) [reply framework/src/utils/Units.C]: Also, why is there no 'minute' unit?

**Comment by andrsd** [comment]
You can ignore the RELAP-7 failure, this PR works on my machine with the latest RELAP-7. So, I will deal with test after this is merged - if it is merged :-)

File: framework/src/utils/Units.C
Comment (dschwen) [reply framework/src/utils/Units.C]: Yeah... because that would create an ambiguity between milli-inch (min) and minute (min).
We could just accept that (as the ambiguous units have different dimensions - length vs. time), but I chose to avoid that for now.

There are potential paths for providing a disambiguation (by permitting an optional separating character between SI prefix and unit, making milli inch `m_in` or `m.in`), but I was reluctant to introduce that complication (at least in the first iteration).

**Comment by dschwen** [comment]
@andrsd can we get this merged with a followup fixup, or would you like (a review-invalidating) patch now?

**Comment by dschwen** [comment]
Ok, some more changes. Still no minutes but you can write `ft/60s` now (to get feet per minute). Also `1/s` is now supported (which previously had to be `s^-1`). Unit tests are added for that new functionality.

File: framework/src/utils/Units.C
Comment (andrsd) [reply framework/src/utils/Units.C]: Why could not you do `min` for minute? Just asking...

Comment (dschwen) [reply framework/src/utils/Units.C]: See above. min could be interpreted as milli inch.

**Comment by andrsd** [comment]
> @andrsd can we get this merged with a followup fixup, or would you like (a review-invalidating) patch now?

Personally, I do not like this approach to doing units - to me this does not really solve anything. If anybody is gonna merge this, then at least fix that comment - which, as I see, you already did.

If you are asking about the RELAP-7 test, I already made some changes: they helped with the convergence, but I need to loosen up some tolerances as well... which I will do sometime today...

File: framework/src/utils/Units.C
Comment (andrsd) [reply framework/src/utils/Units.C]: Got it. I guess I need to wake up first. Thanks!

**Comment by dschwen** [comment]
> Personally, I do not like this approach to doing units

This is not an approach of "doing units" this is providing a unit conversion helper. "Doing units" would either be extremely invasive (using augmented numbers that carry units throughout the code), or put a lot of pressure on the class developers to properly declare input and output units, as well as scaling laws for all quantities!

**Comment by andrsd** [comment]
> This is not an approach of "doing units" this is providing a unit conversion helper.

Fine. Let me rephrase: I do not like this approach of doing unit conversion.

**Comment by dschwen** [comment]
I'm open to alternative suggestions

**Comment by andrsd** [comment]
Here is a few notes on the subject (all are my personal opinions):

1. Units should not be registered globally by framework, but by an application. For example, RELAP-7 does not care about capacitance and we don't want to use `F` for Farad, but for `Fahrenheit`. Framework could provide a database of available units. Maybe, the database should come from a module, so when you used one you would get units defined in it. But, applications should be able to add their own units.

2. Unit should be associated with the input parameter. Example: `addParamWithUnit<Real>("len", 1, Unit::Length, "Doco")`. This should allow for consistency checks, so that if somebody provides a unit of mass, you can tell them to fix their input file. If the unit was requested via this call, I would expect that it should be explicitly listed in the input file, i.e. I would always have to write `len = 1 in` (or whatever length unit). This might be enhanced to allow for default unit - maybe. Depends how draconian you want to be...

3. Input file parser should be enhanced, so that registered units behave like keywords, so when you "parse in" a number you look at the next token and you check if it is a registered unit. If so, you store it with the number so you can do an integrity check for (2). I will admit that I am not sure about `${fparse}`, meaning if this would create bit problems and if units should be allowed inside function expression - or if it is even possible...

I am not much familiar with the parsing system, so I am not sure how difficult (3) is, but in theory it should be doable.  Maybe, you already thought about this and ran into roadblocks and found out this cannot be done this way and you did it with the `${units}` syntax.

**Comment by aeslaughter** [comment]
I have similar feelings as David, but as we go down that road the amount of effort to make this all work is going to be too much. Especially, since where we actually need units is in Material properties and that would be crazy and go back to Cody's comment that would require us to put units in everything.

**Comment by dschwen** [comment]
1. a) No, custom units sound like a bad idea. b) convert 10 'F/m to K/m (good luck!)
2. I have addressed that before (your example is not well thought out as the required automatic scaling needs more than the unit dimension - energy per atom contains length squared but should not necessarily scale with mesh units)
3. That, too. It was the original plan and maybe we'll do it in the future, but it will be very disruptive.

**Comment by dschwen** [comment]
Either way, I'd be happy to replace this PR with one that permits application to add custom brace evalers to the parser. Then I'll just put the unit conversion in marmot or magpie and be done with it. I would very much like to have that convenience for my simulations, but if nobody else needs unit conversion it doesn't have to be in framework.

**Comment by andrsd** [comment]
> 1. No, custom units sound like a bad idea. 

Let's say navier-stokes module has units for mass flow rate. I would like to be able to add a unit like 'mega pound force per hour' which is used in systems code, not may not necessarily be available from the module.

> b) convert 10 'F/m to K/m (good luck!)

You approach is not able to convert Kelvin to Fahrenheit either - if it is what you are suggesting I would have problems with. So, I don't understand your point...

**Comment by andrsd** [comment]
> 2\. your example is not well thought out as the required automatic scaling needs more than the unit dimension - energy per atom contains length squared but should not necessarily scale with mesh units

I was just trying to show a simple example for what I mean by associating the unit with an input parameter. In RELAP-7, users specify length of flow channels, so if they could say 3 feet, instead of 1 meter, they would appreciate it. The second point I was trying to make was that when you associate a unit with an input parameter, you can check for consistency.

How you actually use the parameter internally is a different problem, IMO.

**Comment by rwcarlsen** [comment]
@andrsd - Integrity checks won't work generally unless you flow units through all calculations.  Imagine, for example, a GenericConstantMaterial.  The object itself is unit-agnostic.  So its units should be whatever the user specifies them to be.  So if the user sets it to be `1 ft/s`, then no integrity checking can occur until another objects fetches and uses the material property.  So then we have to flow units with all our numbers - which would be a very big and invasive change.  So associating units with input parameters doesn't actually buy you hardly any more safety - since there would still be many uncheckable cases.  Other than issues with what particular units/conversions are supported, I think this PR provides about as much benefit as can be achieved without drastic mods to moose.

**Comment by rwcarlsen** [comment]
But what about objects that are unit agnostic? - many of them are and should be - like GenericConstantMaterial?  And how will you know what conversion to make? - do you run this check every time we reinit materials?  You can't bind a conversion up front since multiple objects might couple a material but need it in different units from each other.  What you suggest would be much more invasive than you seem to be implying.

**Comment by dschwen** [comment]
Besides Robert's points; You cannot convert derived units that contain Farenheit (or any other affine transformation unit). `Mlb/h` is already a working unit.

**Comment by dschwen** [comment]
> I'd be happy to replace this PR with one that permits application to add custom brace evalers to the parser

I believe 6b51cbd32cb4326df18e7c457ffc797ad996a3f7 would be the change to make that possible (@rwcarlsen ?).

API would be 

```
_parser.addCustomEvaler<UnitConcersionEvaler>("units");
```

(internally uses unique pointers for automatic clean up, `addCustomEvaler` is a variadic template that would accept potential constructor arguments for the evaler, and it builds the object itself to avoid the (IMO awkward) `std::move` stuff to pass in the object.)

**Comment by andrsd** [comment]
@rwcarlsen What I was trying to suggest was that we could help with unit conversion on input parameters in situations like this: Let's say we have an object that takes an input parameter "len" and it expects meters. Now, as a user who wants to use this object I have the dimension in feet, so what I need to do is to convert feet into meters and plug that number into my input file. I would be convenient to be able use "len = x ft" in an input file, i.e. the conversion I am doing "by hand" would be handled by code.

As for `GenericConstantMaterial` - it does not define units via `addParamWithUnits`, so I would not be able to use units with this material's input parameters.

Also, I was not suggesting unit conversion in-between objects, i.e. one object produces values with unit A and object B consumes values with unit B. Users would still need to make sure that units match in this case...

**Comment by dschwen** [comment]
@andrsd nailing objects down to specific units would reduce their flexibility dramatically. MOOSE would not be unit agnostic anymore, you could not use objects over many length and timescales anymore. Think of an object accepting length in meters only (and internally multiplying that length with a gradient), now mesh units would *have* to be in meters to get a coorect result.

**Comment by dschwen** [comment]
This is a bit frustrating for me, as I have been through all this and people kept telling me "units are tricky", "units are harder than you think", yada yada yada. I think I have already realized what the possible complications and barriers here are, and **for that reason** have already backtracked from my original plan, **which was exactly what David is proposing now**.

**Comment by rwcarlsen** [comment]
I think the moose people are generally in support of unit conversion in the input file, but there are a few remaining wishlist items we'd like to see.  @friedmud was saying at lunch that maybe we should use an off-the-shelf unit parsing/conversion package - he had one he was looking at - I'll let him chime in.  Also, we would really want to support `min` as minutes properly (nobody cares about milli-inches) and it really should support Celsius and Fahrenheit conversions - `F` for Fahrenheit would probably get used much more than `F` for Farads too.

**Comment by dschwen** [comment]
Fahrenheit is simply not possible

**Comment by rwcarlsen** [comment]
You can use them for temperature changes just fine.  And when they are combined with other units - that is how they are interpreted.  Just because it is possible for users to do it wrong doesn't mean we shouldn't allow them to do it right.  10F/m is something like "temperature changing 10 degrees F per meter".  Same with C.  One of the most common cases for F and C is to use them as units in isolation - e.g. specifying an inlet temperature or fuel temperature or whatever.  I know it is not convenient in the implementation, but F and C to K and R conversions are likely to be some of the most widely desired/used ones - they should be included here.

**Comment by dschwen** [comment]
Ok, how about we just go with 6b51cbd then?

**Comment by YaqiWang** [comment]
I think the initial design is simple and good, just provide a visualizable way in the input file of how the number is from in a different unit. Whether or not the target unit agrees with the expected unit by the object will be up to users.

**Comment by rwcarlsen** [comment]
I'm not comfortable with encouraging/allowing apps in the MOOSE ecosystem to diverge from each other input-file-syntax-wise.  Maybe it's unavoidable, but it still feels like it's worth working to avoid that whenever possible. And I think we can work out a suitable selection of supported units/conversions here that will be satisfactory for everyone.

**Comment by dschwen** [comment]
> I'm not comfortable with encouraging/allowing apps in the MOOSE ecosystem to diverge from each other input-file-syntax-wise. 

This only superficially makes sense. We have custom actions in apps (which we actually encourage). That's custom syntax that is not much different from adding a new brace expansion keyword within the same syntax framework.

**Comment by joshuahansel** [comment]
My name is Joshua Hansel, and I support this PR. It is simple and effective. Even if we ever wanted to register units, these changes would still be used for that.

My suggestions:
- For the documentation (and this is documentation which people will actually use), I think it would be helpful to organize the quantities by group, i.e., "Pressure", "Length", etc. Also, I think the following looks cleaner, but that's just my opinion (I don't think you need "Description" or to list the SI units like you do in "Base Units"):

## Pressure
Unit | Name | SI Conversion
-|-|-
Pa | Pascal | 1
bar | Bar | 10<sup>5</sup>

If nothing else, remove the extraneous "|"s you have at the end of some rows.

- milliinch can go to hell. How about we restrict the prefixes to SI units? Does anyone ever use prefixes for non-SI units? I say if people want to use **lame** units, they can add "e-3" to their value.

**Comment by joshuahansel** [comment]
What's the plan for temperatures? Fahrenheit and Celsius would be very hot unit conversions. It is unfortunate that it clashes with "Farad" and "Coulomb". Maybe we should require users to use &deg;F and &deg;C 😈

**Comment by lindsayad** [comment]
I think it’s telling that users seem to support this PR. It’s got my vote

**Comment by dschwen** [comment]
If °F and °C and be done without unicode I could add them as special cases where they are the *only* unit (which is the only case that would permit conversion). Let me check. It would also be possible to split the units into two groups. One that supports SI prefixes and one that doesn't. I'd be happy to do that. No SI prefixes for non metric units. I'd still permit them for `bar` and `eV` and those where the prefixes are very commonly used.

**Comment by dschwen** [comment]
`\xc2\xb0` is the 16 bit utf-8 `°` character. This could work (the pre-test wouldn't detect it as unicode in the source - but it might in the test input files)

**Comment by joshuahansel** [comment]
> If °F and °C and be done without unicode I could add them as special cases where they are the only unit (which is the only case that would permit conversion)

I agree that they need to be special cases for when they appear alone since they have the addition component, but isn't the non-alone case even easier, since in that case, we interpret them as temperature *changes*? Specific heat capacity (J/(kg-K)) is the prime example. In that case we have &deg;F=R and &deg;C=K.

**Comment by dschwen** [comment]
> since in that case, we interpret them as temperature changes

Ok, yeah, that would work. I'd still have to think a bit about how to handle the temperature offsets.

**Comment by dschwen** [comment]
:-/ `°` is not permitted in input files. This sucks.Any suggestions? (I don't think ditching SI units is an option). 

```
##########################################################################
ERROR: The following files contain unicode characters
Run: perl -ne 'print if /[^[:ascii:]]/' <filename>
to see what lines are the problem
	test/tests/parser/param_substitution/unit_conversion.i
##########################################################################
```

I totally agree that we should screen for unicode to avoid copy paste mistakes with various dashes that are not `-`, homoglyphs, and invisible composition characters. **But** how about whitelisting a bunch of characters like degree, greek letters, etc.

It would be really nice if the pretest recipe script was actually part of the moose repository, @milljm . That would allow running the pretest locally and to make PRs to the pretest.

**Comment by permcody** [comment]
You pretty much nailed the reasons we screened all Unicode characters back
in the day. If the wider community wants to head the Unicode direction I
won’t hold you back. There may be others that are less thrilled with the
idea. I guess we’ll see ;)

As far as including the pre_check in the repo: The reason has to do with
security. We run those checks for every user, trusted or not. I don’t think
any of us want to execute arbitrary scripts from untrusted sources without
first approving. I can grant you access to change these scripts but I don’t
think we plan to change the location anytime soon.

**Comment by permcody** [comment]
Thinking about this a little more: if customization is what you want
(storing some information about the precheck like whitelist chars) we might
be able to support that as long as inputs as properly sanitized.

**Comment by aeslaughter** [comment]
Can we keep the whitelist for .i files only as well. Currently, MooseDocs doesn't allow unicode and I don't want them getting in until I am ready to support it fully, but I need to think about that for the same reasons. I don't want 100 different quotation and dash styles to start creeping in.

**Comment by dschwen** [comment]
Well you could run the pretest only if the PR does not contain changes to
the pretest scripts OR of the user is trusted...

**Comment by dschwen** [comment]
Ok, I would obviously love to see this go forward. The way to add Celsius and Fahrenheit support that I think is best would involve permitting the degree symbol in the pretest (while that's done please also add greek letters! There should not be any added confusion with those. Please note @aeslaughter that I am not advocating to permit **all** of unicode, just about 20 glyphs).
I don't see why the temperature issue should hold up this PR. This is an incremental improvement that can be done in a follow on PR.

**Comment by joshuahansel** [comment]
> I don't see why the temperature issue should hold up this PR. This is an incremental improvement that can be done in a follow on PR.

> P.S.: If the unicode thing is a no go there could be degF, degC, or oF, oC

I agree, I say let's pick either `degF` or `oF` (my vote is for `degF`, but I'm fine with either). Regardless of whether we allow the degree unicode character eventually, I think it's a good idea to provide a more convenient option than having to use unicode characters.

**Comment by dschwen** [comment]
Ok, I think I got the bugs out. Gotta love unit tests! `degF`, `degC`, and `degR` (Rankine) are implemented now. If they stand by themselves the additive shift is applies (see unit tests). Any binary operator between units or power (applied during parsing of `*`, `/`, `^` of units) will set the additive shift to zero and treat the unit as a *differential*.

P.S.: need to update the docs

**Comment by dschwen** [comment]
Could we please move this along? Temperature units are added @joshuahansel and @rwcarlsen

**Review by lindsayad — APPROVED** [review]
LGTM and it seems like @dschwen has worked hard to address other @idaholab/moose-team concerns

**Comment by joshuahansel** [comment]
I had made some minor documentation suggestions, but they can be implemented later (if at all) - I say this is ready for merge.

**Comment by dschwen** [comment]
@permcody any comments?

**Comment by permcody** [comment]
None

**Comment by fdkong** [comment]
@dschwen Marmot failed with the changes of this PR

https://civet.inl.gov/job/493576/

**Pull request merged**
